sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraIO.java

```diff
@@ -0,0 +1,500 @@
+          .build();
+    }
+
+    public void populateDisplayData(DisplayData.Builder builder) {
+      builder.addIfNotNull(DisplayData.item("hosts", getHosts().toString()));
+      builder.addIfNotNull(DisplayData.item("keyspace", getKeyspace()));
+      builder.add(DisplayData.item("port", getPort()));
+    }
+
+    Cluster getCluster() {
+      LOG.debug("Connecting to Cassandra cluster");
+      return Cluster.builder().addContactPoints(getHosts().toArray(new String[0])).withPort
+          (getPort())
+          .build();
+    }
+  }
+
+  /**
+   * A {@link PTransform} to read data from Apache Cassandra. See {@link CassandraIO} for more
+   * information on usage and configuration.
+   */
+  @AutoValue
+  public abstract static class Read<T> extends PTransform<PBegin, PCollection<T>> {
+
+    @Nullable abstract ConnectionConfiguration getConnectionConfiguration();
+    @Nullable abstract String getQuery();
+    @Nullable abstract String getTable();
+    @Nullable abstract String getRowKey();
+    @Nullable abstract Class<T> getEntityName();
+    @Nullable abstract Coder<T> getCoder();
+
+    abstract Builder<T> builder();
+
+    @AutoValue.Builder
+    abstract static class Builder<T> {
+      abstract Builder<T> setConnectionConfiguration(ConnectionConfiguration
+                                                         connectionConfiguration);
+      abstract Builder<T> setQuery(String query);
+      abstract Builder<T> setTable(String table);
+      abstract Builder<T> setRowKey(String rowKey);
+      abstract Builder<T> setEntityName(Class<T> entityName);
+      abstract Builder<T> setCoder(Coder<T> coder);
+      abstract Read<T> build();
+    }
+
+    public Read<T> withConnectionConfiguration(ConnectionConfiguration connectionConfiguration) {
+      checkArgument(connectionConfiguration != null, "CassandraIO.read()"
+          + ".withConnectionConfiguration(configuration) called with null configuration");
+      return builder().setConnectionConfiguration(connectionConfiguration).build();
+    }
+
+    public Read<T> withQuery(String query) {
+      checkArgument(query != null, "CassandraIO.read().withQuery(query) called with null query");
+      return builder().setQuery(query).build();
+    }
+
+    public Read<T> withTable(String table) {
+      checkArgument(table != null, "CassandraIO.read().withTable(table) called with null table");
+      return builder().setTable(table).build();
+    }
+
+    public Read<T> withRowKey(String rowKey) {
+      checkArgument(rowKey != null, "CassandraIO.read().withRowKey(rowKey) called with null "
+          + "rowKey");
+      return builder().setRowKey(rowKey).build();
+    }
+
+    public Read<T> withEntityName(Class<T> entityName) {
+      checkArgument(entityName != null, "CassandraIO.read().withEntityName(entityName) called "
+          + "with null entityName");
+      return builder().setEntityName(entityName).build();
+    }
+
+    public Read<T> withCoder(Coder<T> coder) {
```

Comment:
AFAIR, I did that at the beginning, but it sounds more convenient to do it in the IO. Let me experiment a bit in the test.